### PR TITLE
Refactor activity result handlers

### PR DESCRIPTION
**Pull request by axuan25**
# Changes

## `onActivityResult`

Previously, SDK integrators had to do their own matching against request codes in `onActivityResult`. Also, when calling `ReachFive.onActivityResult` for SLO flows with a request code that didn't belong to a R5 flow, an exception was thrown.

Things have been simplified:
- Single handler for all login flows: `onLoginActivityResult`.
- If the request code does not concern `ReachFive`, nothing happens (a debug log is emitted to help integrators).
- Helper utilities `isReachFiveLoginRequestCode` & `isReachFiveActionRequestCode` can help integrators determine if the request code belongs to ReachFive.
- `ReachFive.resolveResultHandler` is another helper utility that returns an `ActivityResultHandler` (`LoginResultHandler | WebauthnActionHandler `) which returns a partially applied result handler, integrators only have to pass their success and failure callbacks. 
- Enum class `ErrorCode` helps document error codes.

More details will be included in releases notes & developer documentation.

## `ReachFive` client initialization & `Activity` dependencies

Previously, an `Activity` had to be passed to the `ReachFive` client constructor. This had several negative impacts, the most notable one being that the client had to be instantiated again in each activity, causing repeated network calls for proper configuration initialization.

This dependency was also not uniform with other methods that directly took an `Activity` instance (i.e., SLO flows) instead of using the client's activity. Further, its main use was for loading specific SLO providers (e.g., Google native login) but this makes things unnecessarily complicated for those who don't integrator SLO.

The following simplifications have been performed which should greatly simplify & optimize SDK usage:
- `ReachFive` constructor no longer requires an `Activity`. **The SDK can now be instantiated once for the entire application.**
- `ReachFive.initialize` now only fetches client configuration and no longer loads SLO provider configurations.
- `ReachFive.loadSocialProviders` must now be used to fetch SLO configurations, it takes a `context: Context` so that `Application` can also be passed. The SDK is instantiated and configured only once at most.

### Removed public methods

- `exchangeCodeForToken` is removed as all flow completions are handled by `onLoginActivityResult`.
- `loginCallback` is removed from the public API as it should only be internal.

### CODE REVIEW

File: sdk-core/src/main/java/co/reachfive/identity/sdk/core/ReachFive.kt [hunk 1]
```diff
@@ -3,22 +3,17 @@ package co.reachfive.identity.sdk.core
...
 import co.reachfive.identity.sdk.core.models.ReachFiveError
 import co.reachfive.identity.sdk.core.models.SdkConfig
 import co.reachfive.identity.sdk.core.models.SdkInfos
-import co.reachfive.identity.sdk.core.models.requests.AuthCodeRequest
 import co.reachfive.identity.sdk.core.models.responses.ClientConfigResponse
 import co.reachfive.identity.sdk.core.utils.Failure
 import co.reachfive.identity.sdk.core.utils.Success
 import co.reachfive.identity.sdk.core.utils.SuccessWithNoContent
-
+import java.lang.NullPointerException
```
Comment (FrancoisDevemy5) [inline sdk-core/src/main/java/co/reachfive/identity/sdk/core/ReachFive.kt:16]: Non utilisé

File: sdk-core/src/main/java/co/reachfive/identity/sdk/core/ReachFive.kt [hunk 2]
```diff
@@ -3,22 +3,17 @@ package co.reachfive.identity.sdk.core
...
-import co.reachfive.identity.sdk.core.RedirectionActivity.Companion.ABORT_RESULT_CODE
-import co.reachfive.identity.sdk.core.RedirectionActivity.Companion.CODE_KEY
-import co.reachfive.identity.sdk.core.RedirectionActivity.Companion.CODE_VERIFIER_KEY
-import co.reachfive.identity.sdk.core.RedirectionActivity.Companion.NO_AUTH_ERROR_RESULT_CODE
 import co.reachfive.identity.sdk.core.api.ReachFiveApi
 import co.reachfive.identity.sdk.core.api.ReachFiveApiCallback
 import co.reachfive.identity.sdk.core.models.AuthToken
 import co.reachfive.identity.sdk.core.models.ReachFiveError
 import co.reachfive.identity.sdk.core.models.SdkConfig
 import co.reachfive.identity.sdk.core.models.SdkInfos
```
Comment (FrancoisDevemy5) [inline sdk-core/src/main/java/co/reachfive/identity/sdk/core/ReachFive.kt:11]: Non utilisé

**Review by FrancoisDevemy5 — APPROVED** [review]
J'approuve mais je suis loin de tout comprendre

File: sdk-core/src/main/java/co/reachfive/identity/sdk/core/models/ReachFiveError.kt [hunk 3]
```diff
@@ -59,5 +82,29 @@ data class ReachFiveError(
...
+                    message = apiError.errorDescription ?: "ReachFive API response error",
+                    data = apiError
+                )
+            }
+        }
+
+        @JvmStatic
+        val NoIntent: ReachFiveError =
+            ReachFiveError(
+                code = 52002,
```
[the remark this answers is not in the document]
Comment (nbrr) [reply sdk-core/src/main/java/co/reachfive/identity/sdk/core/models/ReachFiveError.kt:100]: Ce sont des codes custom introduits par cette PR (donc encore non documentés). Par contre ça correspond visiblement à ceux dans `ErrorCode.kt`, qui ne semblent utilisés nulle part.

File: sdk-core/src/main/java/co/reachfive/identity/sdk/core/SessionUtils.kt [hunk 4]
```diff
@@ -28,35 +34,91 @@ internal interface SessionUtils {
...
+        val error = uri.let { ReachFiveError.fromRedirectionResult(it) }
+
+        val authCode = uri.getQueryParameter("code")
+        val codeVerifier = intent?.getStringExtra(CODE_VERIFIER_KEY)
+
+        if (error != null)
+            failure(error)
+        else {
+            if (authCode == null) failure(ReachFiveError.UserCanceled)
+            else if (codeVerifier == null) failure(ReachFiveError.from("Could not retrieve code verifier!"))
```
Comment (nbrr) [inline sdk-core/src/main/java/co/reachfive/identity/sdk/core/SessionUtils.kt:72]: Finalement il reste des erreurs où l'intégrateur ne peut se reposer que sur le message ?

**Review by nbrr — APPROVED** [review]

File: sdk-core/src/main/java/co/reachfive/identity/sdk/core/SessionUtils.kt (hunk 4 above)
Comment (axuan25) [reply sdk-core/src/main/java/co/reachfive/identity/sdk/core/SessionUtils.kt:72]: Oui/non. C'est vraiment une erreur de type unexpected; mais je peux quand lui assigner un code d'erreur tout comme pour l'absence d'`Intent`.

File: sdk-core/src/main/java/co/reachfive/identity/sdk/core/models/ReachFiveError.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (axuan25) [reply sdk-core/src/main/java/co/reachfive/identity/sdk/core/models/ReachFiveError.kt:100]: Merci pour vos retours : il faudrait effectivement utiliser `ErrorCode.NullIntent.code` ici plutôt que l'integer directement.

File: sdk-core/src/main/java/co/reachfive/identity/sdk/core/ActivityResultHandler.kt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import co.reachfive.identity.sdk.core.utils.Success
+
+sealed interface ActivityResultHandler
+
+open class LoginResultHandler(
+    private val reachFive: ReachFive,
+    val requestCode: Int,
+    val resultCode: Int,
+    val intent: Intent?,
+) : ActivityResultHandler {
```
[the remark this answers is not in the document]
Comment (axuan25) [reply sdk-core/src/main/java/co/reachfive/identity/sdk/core/ActivityResultHandler.kt:17]: C'est le style en kotlin / dans le reste du projet.

**Pull request merged**